crates/wallet/src/wallet/persisted.rs

```diff
@@ -1,130 +1,320 @@
-use core::fmt;
+use core::{
+    fmt,
+    future::Future,
+    marker::PhantomData,
+    ops::{Deref, DerefMut},
+    pin::Pin,
+};
 
-use crate::{descriptor::DescriptorError, Wallet};
+use alloc::boxed::Box;
+use chain::Merge;
 
-/// Represents a persisted wallet.
-pub type PersistedWallet = bdk_chain::Persisted<Wallet>;
+use crate::{descriptor::DescriptorError, ChangeSet, CreateParams, LoadParams, Wallet};
 
-#[cfg(feature = "rusqlite")]
-impl<'c> chain::PersistWith<bdk_chain::rusqlite::Transaction<'c>> for Wallet {
-    type CreateParams = crate::CreateParams;
-    type LoadParams = crate::LoadParams;
-
-    type CreateError = CreateWithPersistError<bdk_chain::rusqlite::Error>;
-    type LoadError = LoadWithPersistError<bdk_chain::rusqlite::Error>;
-    type PersistError = bdk_chain::rusqlite::Error;
-
-    fn create(
-        db: &mut bdk_chain::rusqlite::Transaction<'c>,
-        params: Self::CreateParams,
-    ) -> Result<Self, Self::CreateError> {
-        let mut wallet =
-            Self::create_with_params(params).map_err(CreateWithPersistError::Descriptor)?;
-        if let Some(changeset) = wallet.take_staged() {
-            changeset
-                .persist_to_sqlite(db)
+/// Trait that persists [`PersistedWallet`].
+///
+/// For an async version, use [`AsyncWalletPersister`].
+///
+/// Associated functions of this trait should not be called directly, and the trait is designed so
+/// that associated functions are hard to find (since they are not methods!). [`WalletPersister`] is
+/// used by [`PersistedWallet`] (a light wrapper around [`Wallet`]) which enforces some level of
+/// safety. Refer to [`PersistedWallet`] for more about the safety checks.
+pub trait WalletPersister {
+    /// Error type of the persister.
+    type Error;
+
+    /// Initialize the `persister` and load all data.
+    ///
+    /// This is called by [`PersistedWallet::create`] and [`PersistedWallet::load`] to ensure
+    /// the [`WalletPersister`] is initialized and returns all data in the `persister`.
+    ///
+    /// # Implementation Details
+    ///
+    /// The database schema of the `persister` (if any), should be initialized and migrated here.
+    ///
+    /// The implementation must return all data currently stored in the `persister`. If there is no
+    /// data, return an empty changeset (using [`ChangeSet::default()`]).
+    ///
+    /// Error should only occur on database failure. Multiple calls to `initialize` should not
+    /// error. Calling [`persist`] before calling `initialize` should not error either.
```

Comment:
In the docs below for `persist` you said that "The method can fail if the `persister` is not [`initialize`]d" so this comment should be updated to agree right?

```suggestion
    /// error. Calling [`persist`] before calling `initialize` can error.
```

Reply:
Or maybe docs for `persist` should say "This method should ensure the persister is initialized"

Reply:
@ValuedMammal it's a "may" though, since some databases don't require initialization.

```rust
    /// Error should only occur on database failure. Multiple calls to `initialize` should not
    /// error. Calling `initialize` inbetween calls to `persist` should not error.
    ///
    /// Calling [`persist`] before the database is `initialize`d may error. However, some
    /// persister implementations may not require initialization at all (and not error).
```

@notmandatory @ValuedMammal what do you guys thin about this?